web_api/web_api/merge_queues.py

```diff
@@ -54,34 +58,66 @@ def queue_info_from_name(name: str) -> QueueInfo:
     return QueueInfo(org, repo, branch)
 
 
-def get_active_merge_queues(*, install_id: str) -> Mapping[RepositoryName, List[Queue]]:
-    count, keys = r.scan(cursor=0, match=f"merge_queue:{install_id}*", count=50)
+def queue_to_target(queue: bytes) -> bytes:
+    return queue + b":target"
+
+
+def parse_kodiak_queue_entry(data: bytes) -> KodiakQueueEntry | None:
+    try:
+        return KodiakQueueEntry.parse_raw(data)
+    except pydantic.ValidationError:
+        log.exception("failed to parse pull request")
+    return None
+
 
-    # Remove keys for tracking actively merging pull request.
-    #
-    # We append ':target' to the merge queue name to track the merging pull request.
-    # https://github.com/chdsbd/kodiak/blob/b68608e4622ab149997f0cece4d615c2ac51157f/bot/kodiak/queue.py#L41
-    merge_queues = [key for key in keys if not key.endswith(b":target")]
+def get_active_merge_queues(*, install_id: str) -> Mapping[RepositoryName, List[Queue]]:
+    queue_names: Set[bytes] = r.smembers(f"merge_queue_by_install:{install_id}")  # type: ignore [assignment]
     pipe = r.pipeline(transaction=False)
-    for queue in merge_queues:
+    for queue in queue_names:
+        pipe.get(queue_to_target(queue))
+        pipe.get(queue_to_target(queue) + b":time")
         pipe.zrange(queue, 0, 1000, withscores=True)  # type: ignore [no-untyped-call]
+    # response is a list[bytes | None, bytes | None, list[tuple[bytes, float]], ...]
     res = pipe.execute()
 
-    # we accumulate merge queues by repository.
+    it = iter(res)
     queues = defaultdict(list)
-    for queue_name, entries in zip(merge_queues, res):
-        org, repo, branch = queue_info_from_name(queue_name.decode())
-        pull_requests = sorted(
-            [
+    for queue, (merging_pr_raw, current_pr_added_at, waiting_prs) in zip(
+        queue_names, zip(it, it, it)
```

Comment:
lots of zipping

curious what the `it = iter(res)` and the `zip(it, it, it)` does

Reply:
we have a list of results, but we want to turn it into tuples of three items.

```
[a,b,c,a,b,c,a,b,c] => [(a,b,c), (a,b,c),(a,b,c)]
```